src/runner_manager.py

```diff
@@ -0,0 +1,493 @@
+        # setting configured. The is a separated requests Session as the other one configured
+        # according proxies setting provided by user.
+        local_session = requests.Session()
+        local_session.mount("http://", adapter)
+        local_session.mount("https://", adapter)
+
+        self._clients = RunnerClients(
+            GhApi(token=self.config.token),
+            jinja2.Environment(loader=jinja2.FileSystemLoader("templates"), autoescape=True),
+            LxdClient(),
+            RepoPolicyComplianceClient(
+                local_session, "http://127.0.0.1:8080", self.config.service_token
+            ),
+        )
+
+    @retry(tries=5, delay=30, local_logger=logger)
+    def get_latest_runner_bin_url(
+        self, os_name: str = "linux", arch_name: str = "x64"
+    ) -> RunnerApplication:
+        """Get the URL for the latest runner binary.
+
+        The runner binary URL changes when a new version is available.
+
+        Args:
+            os_name: Name of operating system.
+            arch_name: Name of architecture.
+
+        Returns:
+            Information on the runner application.
+        """
+        runner_bins: RunnerApplicationList = []
+        if isinstance(self.config.path, GitHubRepo):
+            runner_bins = self._clients.github.actions.list_runner_applications_for_repo(
+                owner=self.config.path.owner, repo=self.config.path.repo
+            )
+        if isinstance(self.config.path, GitHubOrg):
+            runner_bins = self._clients.github.actions.list_runner_applications_for_org(
+                org=self.config.path.org
+            )
+
+        logger.debug("Response of runner binary list: %s", runner_bins)
+
+        try:
+            return next(
+                bin
+                for bin in runner_bins
+                if bin["os"] == os_name and bin["architecture"] == arch_name
+            )
+        except StopIteration as err:
+            raise RunnerBinaryError(
+                f"Unable query GitHub runner binary information for {os_name} {arch_name}"
+            ) from err
+
+    @retry(tries=5, delay=30, local_logger=logger)
+    def update_runner_bin(self, binary: RunnerApplication) -> None:
+        """Download a runner file, replacing the current copy.
+
+        Remove the existing runner binary to prevent it from being used. This
+        is done to prevent security issues arising from outdated runner binary
+        containing security flaws. The newest version of runner binary should
+        always be used.
+
+        Args:
+            binary: Information on the runner binary to download.
+        """
+        logger.info("Downloading runner binary from: %s", binary["download_url"])
+
+        # Delete old version of runner binary.
+        RunnerManager.runner_bin_path.unlink(missing_ok=True)
+
+        # Download the new file
+        response = self.session.get(binary["download_url"], stream=True)
```

Comment:
It would be good to handle the exceptions when accessing other systems via unreliable channels like the network. Unexpected scenarios are likely to arise. Same for the rest of the scenarios